### PR TITLE
Improved output/workflow of 'verify' (including 'pull')

### LINKED ISSUE #3646: Pull signed images from libraries returns non-zero exit status
## Version of Singularity:

Master branch (build: 5fac9d2).

### Expected behavior

Pulling signed images from libraries should return 0 for exit status if no errors occurred.

### Actual behavior

Pulling signed images from libraries returns 1 for exit status.

### Steps to reproduce behavior

`$ singularity pull library://sylabs/tests/signed:1.0.0`
INFO:    Key with ID EDECE4F3F38D871E not found in local keyring, downloading from keystore...
INFO:    Container is signed
Data integrity checked, authentic and signed by:
        Sylabs Admin <support@sylabs.io>, Fingerprint 8883491F4268F173C6E5DC49EDECE4F3F38D871E
WARNING: no local key matching entity
INFO:    Download complete: signed_1.0.0.sif
`$ echo $?`
1

### PR DESCRIPTION
**Pull request by WestleyK**
## Description of the Pull Request (PR):

This PR improves the output of `verify` when verifying a container. Suggested by @gmkurtzer.

### Example:

```bash
$ singularity pull library://westleyk/demo/signed-test
INFO:    Downloading library image
 2.61 MiB / 2.61 MiB [===========================================================================================] 100.00% 4.23 MiB/s 0s
ERROR:   unable to verify container: signed-test_latest.sif
This image is not signed, and thus its contents cannot be verified.

$ singularity verify signed-test_latest.sif 
Container is signed by 6 key(s):

Verifying signature F: 8883491F4268F173C6E5DC49EDECE4F3F38D871E:
[LOCAL]    Sylabs Admin <support@sylabs.io>

Verifying signature F: 36E5D93B03B37178CDC9E9D7AA246EFBB99BBE05:
[MISSING]  key does not exist in local, or remote keystore

Verifying signature F: F69C21F759C8EA06FD32CCF4536523CE1E109AF3:
[LOCAL]    E2E test key (E2E test key; do not use) <westley@sylabs.io>

Verifying signature F: C7E7C8C3635DD06930669A2283B30190FEEF8162:
[LOCAL]    Westley (from gpg) <westley@sylabs.io>

Verifying signature F: E09D378BD534368E8BF2390386216BF0A30345A0:
[REMOTE]   WestleyK (test key 1) <westley@sylabs.io>

Verifying signature F: C7E7C8C3635DD06930669A2283B30190FEEF8162:
[LOCAL]    Westley (from gpg) <westley@sylabs.io>

FATAL:   Failed to verify: signed-test_latest.sif: unble to fetch key 36E5D93B03B37178CDC9E9D7AA246EFBB99BBE05: no matching keys found for fingerprint
```

And a more common alpine container:

```bash
$ singularity pull library://alpine
INFO:    Downloading library image
 2.59 MiB / 2.59 MiB [==============================================================] 100.00% 3.43 MiB/s 0s
INFO:    Container is trusted - run 'singularity key list' to list your trusted keys
INFO:    Download complete: alpine_latest.sif

$ singularity verify alpine_latest.sif 
Container is signed by 1 key(s):

Verifying signature F: 8883491F4268F173C6E5DC49EDECE4F3F38D871E:
[LOCAL]    Sylabs Admin <support@sylabs.io>

Signature(s) and data integrity verified
```

_And of course, there will also be some color in the example, but its not shown here._

## This fixes or addresses the following GitHub issues:

- Fixes #3646 

<br>

### CODE REVIEW

File: pkg/signing/signing.go [hunk 1]
```diff
@@ -248,6 +248,7 @@ func IsSigned(cpath, keyServerURI string, id uint32, isGroup bool, authToken str
 	}
 	if noLocalKey {
 		return true, fmt.Errorf("no local key matching entity")
+		//return true, nil
```
Comment (jchavezlavalle) [inline pkg/signing/signing.go]: remove this line

File: pkg/signing/signing.go [hunk 2]
```diff
@@ -308,18 +311,23 @@ func Verify(cpath, keyServiceURI string, id uint32, isGroup bool, authToken stri
 		}
 
 		// verify the container with our local keys first
+		sylog.Verbosef("Container signature found: %s\n", fingerprint)
 		signer, err := openpgp.CheckDetachedSignature(elist, bytes.NewBuffer(block.Bytes), block.ArmoredSignature.Body)
 		if err != nil {
+			trusted = false
+			notLocalKey = true
+
 			// if theres a error, thats proboly becuse we dont have a local key
```
Comment (jchavezlavalle) [inline pkg/signing/signing.go]: check typos on comments

File: pkg/signing/signing.go [hunk 3]
```diff
@@ -280,6 +281,8 @@ func Verify(cpath, keyServiceURI string, id uint32, isGroup bool, authToken stri
 
 	// compare freshly computed hash with hashes stored in signatures block(s)
 	for _, v := range signatures {
+		trusted := true
+
```
Comment (jchavezlavalle) [inline pkg/signing/signing.go]: remove additional lines

**Review by jchavezlavalle — CHANGES_REQUESTED** [review]

**Comment by WestleyK** [comment]
@jscook2345, @devprincess, @tri-adam, @GodloveD, @ikaneshiro, @rherban,
Do you have any thoughts on the output?

```bash
$ singularity pull library://westleyk/demo/signed-test
[...]
$ singularity verify signed-test_latest.sif 
ERROR:   Could not obtain key from remote keystore: B99BBE05: no matching keys found for fingerprint
INFO:    Found key in remote keystore: 1E109AF3
INFO:    Found key in remote keystore: A30345A0

Data integrity checked, authentic and signed by:
[TRUSTED]  Sylabs Admin <support@sylabs.io>, F: 8883491F4268F173C6E5DC49EDECE4F3F38D871E
[MISSING]  key does not exist in local, or remote keystore: 36E5D93B03B37178CDC9E9D7AA246EFBB99BBE05
           E2E test key (E2E test key; do not use) <westley@sylabs.io>, F: F69C21F759C8EA06FD32CCF4536523CE1E109AF3
[TRUSTED]  Westley (from gpg) <westley@sylabs.io>, F: C7E7C8C3635DD06930669A2283B30190FEEF8162
           WestleyK (test key 1) <westley@sylabs.io>, F: E09D378BD534368E8BF2390386216BF0A30345A0
[TRUSTED]  Westley (from gpg) <westley@sylabs.io>, F: C7E7C8C3635DD06930669A2283B30190FEEF8162
```

_EDIT: this output is outdated now._

**Comment by ikaneshiro** [comment]
@WestleyK couple questions related to the example output:
1. This image appears to be signed by multiple signatures? I see `B99BBE05`, `1E109AF3` and `A30345A0` in the error and info messages, but the data integrity check lists five unique fingerprints in its output. Where are the two additional fingerprints from?
2. Why is this line printed twice? `[TRUSTED]  Westley (from gpg) <westley@sylabs.io>, F: C7E7C8C3635DD06930669A2283B30190FEEF8162`

**Comment by WestleyK** [comment]
@ikaneshiro,

> 1. This image appears to be signed by multiple signatures? I see `B99BBE05`, `1E109AF3` and `A30345A0` in the error and info messages, but the data integrity check lists five unique fingerprints in its output. Where are the two additional fingerprints from?

The `INFO:` is saying "you dont have a local key, but theres one in the keystore" (untrusted key). The `ERROR:` is saying that the key (`B99BBE05`) does not exist in the keystore, or local keyring (a revoked key, or a unpushed key). All the other keys that signed that container are trusted, (I have them locally).

<br>

>Why is this line printed twice? `[TRUSTED] Westley (from gpg) <westley@sylabs.io>, F: C7E7C8C3635DD06930669A2283B30190FEEF8162`

Because I signed that container twice with that key :smile:

File: cmd/singularity/pull_test.go [hunk 4]
```diff
@@ -56,16 +50,20 @@ func imagePull(t *testing.T, library, pullDir string, imagePath string, sourceSp
 	return cmd.CombinedOutput()
 }
 
+// tmpDirReturn will return a tmp dir path in /tmp.
+func tmpDirReturn(t *testing.T) string {
```
Comment (ikaneshiro) [inline cmd/singularity/pull_test.go]: Can we change this to `makeTmpDir` or something to that effect? Having `Return` in the name isn't very descriptive.

File: go.mod [hunk 5]
```diff
@@ -34,7 +34,7 @@ require (
 	github.com/docker/go-units v0.3.3 // indirect
 	github.com/docker/libtrust v0.0.0-20160708172513-aabc10ec26b7 // indirect
 	github.com/dsnet/compress v0.0.1 // indirect
-	github.com/fatih/color v1.7.0 // indirect
+	github.com/fatih/color v1.7.1-0.20181010231311-3f9d52f7176a
```
Comment (ikaneshiro) [inline go.mod]: Is there a reason you're using a specific commit of this package instead of using a tagged version?

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
Couple nit picks but I think this is well on its way. Also, can you add some sort of success message when container verification succeeds? like `INFO: Container signatures and data integrity verified`

File: go.mod (hunk 5 above)
Comment (WestleyK) [reply go.mod]: Umm, I guess this vendor was added after I made my PR (it was not there a couple weeks ago, or I probably did not see it)... I'll change it back.

File: cmd/singularity/pull_test.go (hunk 4 above)
Comment (WestleyK) [reply cmd/singularity/pull_test.go]: Actually, this should not even be in this PR, I'll make a new PR that improves the `pull_test.go`

Comment (WestleyK) [reply cmd/singularity/pull_test.go]: Nevermined, these changes need to be in this PR (this PR changes the exit code).

**Comment by WestleyK** [comment]
Closing. https://github.com/sylabs/singularity/pull/3896 is cleaner (less commits)

**Pull request closed without merging**